File: 2023/Ruby/day1.rb
file = File.open('../inputs/day1.txt')
text = file.read

total = 0

for line in text.lines do
  numbers = Array.new
  number = 0

  line.each_char do |character|
    begin
      number = Integer(character)
      numbers.push(number)
    rescue ArgumentError
    end
  end

  if numbers.length >= 2
    number = "#{numbers[0]}#{numbers[numbers.length - 1]}".to_i
  elsif numbers.length == 1
    number = "#{numbers[0]}#{numbers[0]}".to_i
  else
    number = 0
  end

  total += number
end

puts total

File: 2023/Lua/day1.lua
-- Lua 5.4.6

function openfile(filepath)
    local file = io.open(filepath)

    if file == nil then
        local error_message = string.format("cannot open file at path \"%s\"", filepath)
        error(error_message)
    end

    return file
end


local filepath = "../inputs/day1.txt"
local success, file = pcall(openfile, filepath)
Total = 0

if success then
    for line in file:lines() do
        local numbers = {}
    
        for i = 1, #line do
            local character = string.sub(line, i, i)
    
            if tonumber(character) then
                table.insert(numbers, character)
            end
        end
    
        -- I often forgot that array index in Lua start at 1
    
        if #numbers >= 2 then
            local number = numbers[1] .. numbers[#numbers]
            Total = Total + tonumber(number)
        elseif #numbers == 1 then
            local number = numbers[1] .. numbers[1]
            Total = Total + tonumber(number)
        end
    end
    
    print(Total)

else
    print("Error:", file)
end